### PR TITLE
Support scala-natve 0.5.0

### PR DESCRIPTION
**Pull request by catap**
It has a few changes in build plugin, which requires to cleanup a bit build script.

And one deprectated class was removed.

Tested as:
```
export SCALANATIVE_VERSION=0.5.0-RC2

sbt clean

sbt ++2.12.17 "project scalacticNative" clean publishLocal
sbt ++2.13.10 "project scalacticNative" clean publishLocal
sbt "project scalacticDottyNative" clean publishLocal

sbt ++2.12.17 "project scalatestNative" clean publishLocal
sbt ++2.13.10 "project scalatestNative" clean publishLocal
sbt "project scalatestDottyNative" clean publishLocal

sbt ++2.12.17 "project scalatestAppNative" clean publishLocal
sbt ++2.13.10 "project scalatestAppNative" clean publishLocal
sbt "project scalatestAppDottyNative" clean publishLocal
```

### CODE REVIEW

**Comment by catap** [comment]
BTW to build it you must to publish localy `scala-xml` for the same version of Scala-Native.

**Comment by cheeseng** [comment]
@catap I am trying this, seems like when building with Scala 3 it will fail with: 

```
[error] sbt.librarymanagement.ResolveException: Error downloading org.scala-lang.modules:scala-xml_native0.5.0-RC2_3:2.1.0
[error]   Not found
[error]   Not found
[error]   not found: /home/cheeseng/.ivy2/local/org.scala-lang.modules/scala-xml_native0.5.0-RC2_3/2.1.0/ivys/ivy.xml
[error]   not found: https://repo1.maven.org/maven2/org/scala-lang/modules/scala-xml_native0.5.0-RC2_3/2.1.0/scala-xml_native0.5.0-RC2_3-2.1.0.pom
[error]   not found: https://jcenter.bintray.com/org/scala-lang/modules/scala-xml_native0.5.0-RC2_3/2.1.0/scala-xml_native0.5.0-RC2_3-2.1.0.pom
[error]   not found: https://oss.sonatype.org/content/groups/public/org/scala-lang/modules/scala-xml_native0.5.0-RC2_3/2.1.0/scala-xml_native0.5.0-RC2_3-2.1.0.pom
[error]   not found: /home/cheeseng/.m2/repository/org/scala-lang/modules/scala-xml_native0.5.0-RC2_3/2.1.0/scala-xml_native0.5.0-RC2_3-2.1.0.pom
[error]   not found: https://oss.sonatype.org/content/repositories/orgscalactic-1461/org/scala-lang/modules/scala-xml_native0.5.0-RC2_3/2.1.0/scala-xml_native0.5.0-RC2_3-2.1.0.pom
[error]   not found: https://oss.sonatype.org/content/repositories/orgscalatest-1754/org/scala-lang/modules/scala-xml_native0.5.0-RC2_3/2.1.0/scala-xml_native0.5.0-RC2_3-2.1.0.pom
[error] Error downloading org.scala-native:test-interface_native0.4_3:0.5.0-RC2
[error]   Not found
[error]   Not found
[error]   not found: /home/cheeseng/.ivy2/local/org.scala-native/test-interface_native0.4_3/0.5.0-RC2/ivys/ivy.xml
[error]   not found: https://repo1.maven.org/maven2/org/scala-native/test-interface_native0.4_3/0.5.0-RC2/test-interface_native0.4_3-0.5.0-RC2.pom
[error]   not found: https://jcenter.bintray.com/org/scala-native/test-interface_native0.4_3/0.5.0-RC2/test-interface_native0.4_3-0.5.0-RC2.pom
[error]   not found: https://oss.sonatype.org/content/groups/public/org/scala-native/test-interface_native0.4_3/0.5.0-RC2/test-interface_native0.4_3-0.5.0-RC2.pom
[error]   not found: /home/cheeseng/.m2/repository/org/scala-native/test-interface_native0.4_3/0.5.0-RC2/test-interface_native0.4_3-0.5.0-RC2.pom
[error]   not found: https://oss.sonatype.org/content/repositories/orgscalactic-1461/org/scala-native/test-interface_native0.4_3/0.5.0-RC2/test-interface_native0.4_3-0.5.0-RC2.pom
[error]   not found: https://oss.sonatype.org/content/repositories/orgscalatest-1754/org/scala-native/test-interface_native0.4_3/0.5.0-RC2/test-interface_native0.4_3-0.5.0-RC2.pom
```

Perhaps the related scala-xml is not published yet?

**Comment by catap** [comment]
@cheeseng ineed, I've locally updated to scala-xml-2.2.0, and publish it for Scala-Native-0.5.0-RC2 to unblock it.

Unfortently my yesterday test wasn't full, and required changed leads to update of scala3 version as well (otherwise it will be missmatch of ABI verison), and the last leads to a wired error:

```
[error] -- Error: /home/catap/src/scalatest/dotty/core.native/target/scala-3/src_managed/main/org/scalatest/Doc.scala:112:86 
[error] 112 |  final override lazy val nestedSuites: collection.immutable.IndexedSeq[Suite] = for (IncludedSuite(suite) <- snippets) yield suite
[error]     |                                                                                      ^^^^^^^^^^^^^^^^^^^^
[error]     |pattern's type Doc.this.IncludedSuite is more specialized than the right hand side expression's type Doc.this.Snippet
[error]     |
[error]     |If the narrowing is intentional, this can be communicated by adding the `case` keyword before the full pattern,
[error]     |which will result in a filtering for expression (using `withFilter`).
[error]     |This patch can be rewritten automatically under -rewrite -source 3.2-migration.
```

So, if you have any advice how to fix it, I'd like any help. Or, feel free to push into my branch.

My goal is prepare scalatest for the future release of scala-native, at least locally, to run heavy unit test on my code base, which usually allows to catch wired bugs :)

**Comment by catap** [comment]
Well, seems that downgrade to `scala-3.3.3` avoid to compile everything without this wired issue.

I continue to test it, and as soon as it's fully tested will convert this draft back to PR.

**Comment by catap** [comment]
BTW opened a question in scala-xml: https://github.com/scala/scala-xml/issues/713

**Comment by cheeseng** [comment]
@catap Fyi to build with Scala 3.4 you'll need: 

https://github.com/scalatest/scalatest/compare/main...cheeseng:scalatest:feature-scala-3.4

But I think Scala 3.3 should be the better target since it is LTS.

**Comment by catap** [comment]
@cheeseng everything had passed with scala-3.3.3

**Comment by catap** [comment]
@cheeseng you seems the right person to ask. During my test migration a few tests are failed with reason:
```
java.lang.IllegalStateException: Queue is empty while future is not completed, this means you're probably using a wrong ExecutionContext for your task, please double check your Future
```
can you suggest any easy way to unblock it?

**Comment by catap** [comment]
And...
```
  implicit override def executionContext: ExecutionContext =
    ExecutionContext.global
```
seems to fix this test

**Comment by catap** [comment]
After future digging, just thoughts.

Probably https://github.com/scalatest/scalatest/pull/1858 need to be "adjusted". It doesn't requires for Scala-Native-0.5.0 anymore when Multithreading is enabled.

It can be checked via `scala.scalanative.meta.LinktimeInfo.isMultithreadingEnabled` which is linking-time condition and allows to have different code when multithreading is enabled and disabled.

But it exists on 0.5.0 only, so everything before will be literally broken.

I have no idea how to support old scala native and new one on the same code base.

**Comment by cheeseng** [comment]
>> I have no idea how to support old scala native and new one on the same code base.

@catap Hmm, I wonder if we still want to support scala-native 0.4 going forward?  If need to we probably can do so with a generator in sbt build I think.

**Comment by catap** [comment]
On Sun, 17 Mar 2024 01:07:10 +0100,
Chua Chee Seng wrote:
> 
> @catap Hmm, I wonder if we still want to support scala-native 0.4 going forward?  If need to we probably can do so with a generator in sbt build I think.
> 

Frankly speaking I had thought the same. Scala-Native 0.5.0 brings support
of multithreading which is a kind of deal breaker.

As far as I can say the only platform where it doesn't work which SN
supports is NetBSD. Why? Because I haven't found enough time to polish it,
but I absolutley sure that I'll do it sooner or later.

And I assume that set of users of Scala-Native on NetBSD is empty because
I've added experemental support a week ago, and such use case can be
ignored.

So, where I stay it seems that droping support of scala-native before 0.5.0
at the next release and make things easy to support... well, I haven't see
why not.

Special that SN-0.5.0 doesn't binary compatible with old version and
requires to republish everything.

Or, add some macros like SCALA-NATIVE-BEFORE-0.5.0 but when it should end?

-- 
wbr, Kirill

**Comment by catap** [comment]
@cheeseng I just added one more commit which rework https://github.com/scalatest/scalatest/pull/1858 and brokes compatibillity with Scala-Native before 0.5.0

Trivial test case work on both enabled and disabled multithreading builds.

The whole test case almost work, but it may stuck on some random test. Probably related to issues inside SN itself, see for example: https://github.com/scala-native/scala-native/issues/3846

File: jvm/featurespec-test/src/test/scala/org/scalatest/featurespec/AsyncFeatureSpecLikeSpec.scala [hunk 1]
```diff
@@ -44,6 +44,10 @@ class AsyncFeatureSpecLikeSpec extends funspec.AnyFunSpec {
       class ExampleSpec extends AsyncFeatureSpecLike with ParallelTestExecution {
 
         //SCALATESTJS-ONLY implicit override def executionContext = org.scalatest.concurrent.TestExecutionContext.runNow
+        //SCALATESTNATIVE-ONLY implicit override def executionContext =
+        //SCALATESTNATIVE-ONLY   if (scala.scalanative.meta.LinktimeInfo.isMultithreadingEnabled)
+        //SCALATESTNATIVE-ONLY     scala.concurrent.ExecutionContext.Implicits.global
+        //SCALATESTNATIVE-ONLY   else super.executionContext
```
Comment (WojciechMazur) [inline jvm/featurespec-test/src/test/scala/org/scalatest/featurespec/AsyncFeatureSpecLikeSpec.scala:50]: Is it really needed? Scala Native does already use real, ForkJoinPool based ExecutionContext when multithreading is enabled or QueueExecutionContext in singlet-hreaded mode. Although I don't know what lives under `super.executionContext` 
https://github.com/scala-native/scala-native/blob/8ac5d27670389495ca3493213a61105df9a6f95b/scalalib/overrides-2.13/scala/concurrent/ExecutionContext.scala.patch#L16

Comment (catap) [reply jvm/featurespec-test/src/test/scala/org/scalatest/featurespec/AsyncFeatureSpecLikeSpec.scala:50]: @WojciechMazur thanks fro you feedback. It uses https://github.com/scalatest/scalatest/blob/3.2.x-new/jvm/core/src/main/scala/org/scalatest/concurrent/SerialExecutionContext.scala under `super.executionContext`.

Comment (catap) [reply jvm/featurespec-test/src/test/scala/org/scalatest/featurespec/AsyncFeatureSpecLikeSpec.scala:50]: Long story short: with this version it work with and without multithreading out the box.

I had encountered two issues https://github.com/scala-native/scala-native/issues/3846 and https://github.com/scala-native/scala-native/issues/3848 with enabled multithreading, but it seems to be Scala-Native related or/and my code (libuv integration) which I still "clean" for multithreading.

**Comment by WojciechMazur** [comment]
I think this PR still is not ready and needs adaption to added multithreading support in Scala Native. 
Theres a plenty of tests that fail especially the async one that were referenced in reported SN issue. 

Test command: `scalatestFeatureSpecTestDottyNative/test`
Output: 
```scala
[error] Failed tests:
[error]         org.scalatest.featurespec.DeprecatedAsyncFeatureSpecSpec
[error]         org.scalatest.featurespec.AsyncFeatureSpecLikeSpec
[error]         org.scalatest.featurespec.FixtureAsyncFeatureSpecLikeSpec
[error]         org.scalatest.featurespec.DeprecatedAsyncFeatureSpecLikeSpec
[error]         org.scalatest.featurespec.AsyncFeatureSpecSpec
[error]         org.scalatest.featurespec.FixtureAsyncFeatureSpecSpec
```

**Comment by catap** [comment]
On Sun, 24 Mar 2024 23:38:53 +0100,
Wojciech Mazur wrote:
> 
> I think this PR still is not ready and needs adaption to added multithreading support in Scala Native. 
> Theres a plenty of tests that fail especially the async one that were referenced in reported SN issue. 
> 
> Test command: `scalatestFeatureSpecTestDottyNative/test`
> Output: 
> ```scala
> [error] Failed tests:
> [error]         org.scalatest.featurespec.DeprecatedAsyncFeatureSpecSpec
> [error]         org.scalatest.featurespec.AsyncFeatureSpecLikeSpec
> [error]         org.scalatest.featurespec.FixtureAsyncFeatureSpecLikeSpec
> [error]         org.scalatest.featurespec.DeprecatedAsyncFeatureSpecLikeSpec
> [error]         org.scalatest.featurespec.AsyncFeatureSpecSpec
> [error]         org.scalatest.featurespec.FixtureAsyncFeatureSpecSpec
> ```
> 

Good point. I'll dig into this and do somethig with all these tests.

Maybe it allows to reproduce the deadlock.

-- 
wbr, Kirill

**Comment by catap** [comment]
@cheeseng, @WojciechMazur I've removed the last commit. Without it scalatests' unit test passes on 0.5.0-RC2, it keeps code cleanaer, and it hasn't got any effect on deadlock inside tests on my code.

**Comment by catap** [comment]
I've simplified my test to clean reproducer which I can use to explain a kind of deadlock which I able to overstep by
```
  implicit override def executionContext: ExecutionContext = ExecutionContext.global
```
So, the test looks like:
```
    class SomeError() extends Exception
    for {
      promise1 <- timedPromise[Unit](20.millis) { p =>
        p.trySuccess(())
      }
      promise2 <- timedPromise[Unit](10.millis) { p =>
        promise1.tryFailure(new SomeError())
        p.trySuccess(())
      }
      _ <- promise2.future
      _ <- recoverToSucceededIf[SomeError](promise1.future)
    } yield Succeeded
```
where `timedPromise` create a timer via `uv` and executes a task which run `uv_run(uvLoop, uv_run_mode.DEFAULT)` inside `implicit ec: ExecutionContext`.

If I don't hardcoded inside test to use `ExecutionContext.global` test is blocked inside `SerialExecutionContext` which is blocked inside `uv_run`.

If I hardcode to use `ExecutionContext.global` to execute the libuv task, the code fails as:
```
[info]   java.lang.IllegalStateException: Queue is empty while future is not completed, this means you're probably using a wrong ExecutionContext for your task, please double check your Future.
[info]   at java.lang.Throwable.fillInStackTrace(Unknown Source)
[info]   at org.scalatest.concurrent.SerialExecutionContext.recRunNow(Unknown Source)
[info]   at org.scalatest.AsyncSuperEngine.runTestImpl(Unknown Source)
[info]   at org.scalatest.wordspec.AsyncWordSpecLike.runTest(Unknown Source)
[info]   at org.scalatest.wordspec.AsyncWordSpecLike$$Lambda$20.apply(Unknown Source)
[info]   at org.scalatest.AsyncSuperEngine.traverseSubNodes$1$$anonfun$1(Unknown Source)
[info]   at org.scalatest.AsyncSuperEngine$$Lambda$22.apply(Unknown Source)
[info]   at scala.collection.LinearSeqOps.foldLeft(Unknown Source)
[info]   at org.scalatest.AsyncSuperEngine.traverseSubNodes$1(Unknown Source)
[info]   at org.scalatest.AsyncSuperEngine.runTestsInBranch(Unknown Source)
```

I may reorganize the code and make dedicated function to run event loop, but it seems a bit wired in case of multithreading, isn't it?

**Comment by lbialy** [comment]
@catap can we somehow help with this?

**Comment by catap** [comment]
On Fri, 12 Apr 2024 15:08:14 +0200,
Łukasz Biały wrote:
> 
> @catap can we somehow help with this? 
> 

As soon as scala-xml and other dependencies are released for SN-0.5.0, I'll
update this branch.

Right now it works well with one deadlock which can be reproduced by
specific edge case (my codebase), and I think it should be fixed inside SN,
and not scalatests.

So, I live with assumption that as soon as dependencies are updated, it
should pass CI and can be merged and released.

-- 
wbr, Kirill

**Comment by cheeseng** [comment]
@catap I just tried with the released 0.5.0, seems working fine for Scala 2.12 and 2.13 (Scala 2.11 seems not supported), but for the scala-xml 2.2.0 the build for Scala 3 seems like not available yet: 

```
[error] sbt.librarymanagement.ResolveException: Error downloading org.scala-lang.modules:scala-xml_native0.5_3:2.2.0
[error]   Not found
[error]   Not found
[error]   not found: /home/cheeseng/.ivy2/local/org.scala-lang.modules/scala-xml_native0.5_3/2.2.0/ivys/ivy.xml
[error]   not found: https://repo1.maven.org/maven2/org/scala-lang/modules/scala-xml_native0.5_3/2.2.0/scala-xml_native0.5_3-2.2.0.pom
[error]   not found: https://jcenter.bintray.com/org/scala-lang/modules/scala-xml_native0.5_3/2.2.0/scala-xml_native0.5_3-2.2.0.pom
[error]   not found: https://oss.sonatype.org/content/groups/public/org/scala-lang/modules/scala-xml_native0.5_3/2.2.0/scala-xml_native0.5_3-2.2.0.pom
[error]   not found: /home/cheeseng/.m2/repository/org/scala-lang/modules/scala-xml_native0.5_3/2.2.0/scala-xml_native0.5_3-2.2.0.pom
[error]   not found: https://oss.sonatype.org/content/repositories/orgscalactic-1461/org/scala-lang/modules/scala-xml_native0.5_3/2.2.0/scala-xml_native0.5_3-2.2.0.pom
[error]   not found: https://oss.sonatype.org/content/repositories/orgscalatest-1754/org/scala-lang/modules/scala-xml_native0.5_3/2.2.0/scala-xml_native0.5_3-2.2.0.pom
[error] 	at lmcoursier.CoursierDependencyResolution.unresolvedWarningOrThrow(CoursierDependencyResolution.scala:258)
[error] 	at lmcoursier.CoursierDependencyResolution.$anonfun$update$38(CoursierDependencyResolution.scala:227)
[error] 	at scala.util.Either$LeftProjection.map(Either.scala:573)
[error] 	at lmcoursier.CoursierDependencyResolution.update(CoursierDependencyResolution.scala:227)
[error] 	at sbt.librarymanagement.DependencyResolution.update(DependencyResolution.scala:60)
[error] 	at sbt.internal.LibraryManagement$.resolve$1(LibraryManagement.scala:53)
[error] 	at sbt.internal.LibraryManagement$.$anonfun$cachedUpdate$12(LibraryManagement.scala:103)
[error] 	at sbt.util.Tracked$.$anonfun$lastOutput$1(Tracked.scala:73)
[error] 	at sbt.internal.LibraryManagement$.$anonfun$cachedUpdate$20(LibraryManagement.scala:116)
[error] 	at scala.util.control.Exception$Catch.apply(Exception.scala:228)
[error] 	at sbt.internal.LibraryManagement$.$anonfun$cachedUpdate$11(LibraryManagement.scala:116)
[error] 	at sbt.internal.LibraryManagement$.$anonfun$cachedUpdate$11$adapted(LibraryManagement.scala:97)
[error] 	at sbt.util.Tracked$.$anonfun$inputChangedW$1(Tracked.scala:219)
[error] 	at sbt.internal.LibraryManagement$.cachedUpdate(LibraryManagement.scala:130)
[error] 	at sbt.Classpaths$.$anonfun$updateTask0$5(Defaults.scala:3525)
[error] 	at scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] 	at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
[error] 	at sbt.std.Transform$$anon$4.work(Transform.scala:68)
[error] 	at sbt.Execute.$anonfun$submit$2(Execute.scala:282)
[error] 	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:23)
[error] 	at sbt.Execute.work(Execute.scala:291)
[error] 	at sbt.Execute.$anonfun$submit$1(Execute.scala:282)
[error] 	at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
[error] 	at sbt.CompletionService$$anon$2.call(CompletionService.scala:64)
[error] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[error] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[error] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[error] 	at java.lang.Thread.run(Thread.java:750)
[error] (scalatestCoreDottyNative / update) sbt.librarymanagement.ResolveException: Error downloading org.scala-lang.modules:scala-xml_native0.5_3:2.2.0
[error]   Not found
[error]   Not found
[error]   not found: /home/cheeseng/.ivy2/local/org.scala-lang.modules/scala-xml_native0.5_3/2.2.0/ivys/ivy.xml
[error]   not found: https://repo1.maven.org/maven2/org/scala-lang/modules/scala-xml_native0.5_3/2.2.0/scala-xml_native0.5_3-2.2.0.pom
[error]   not found: https://jcenter.bintray.com/org/scala-lang/modules/scala-xml_native0.5_3/2.2.0/scala-xml_native0.5_3-2.2.0.pom
[error]   not found: https://oss.sonatype.org/content/groups/public/org/scala-lang/modules/scala-xml_native0.5_3/2.2.0/scala-xml_native0.5_3-2.2.0.pom
[error]   not found: /home/cheeseng/.m2/repository/org/scala-lang/modules/scala-xml_native0.5_3/2.2.0/scala-xml_native0.5_3-2.2.0.pom
[error]   not found: https://oss.sonatype.org/content/repositories/orgscalactic-1461/org/scala-lang/modules/scala-xml_native0.5_3/2.2.0/scala-xml_native0.5_3-2.2.0.pom
[error]   not found: https://oss.sonatype.org/content/repositories/orgscalatest-1754/org/scala-lang/modules/scala-xml_native0.5_3/2.2.0/scala-xml_native0.5_3-2.2.0.pom
[error] Total time: 13 s, completed Apr 15, 2024 9:42:02 AM
```

**Comment by catap** [comment]
Now it's blocked by https://github.com/scala/scala-parser-combinators

**Comment by catap** [comment]
Linked issue: https://github.com/scala/scala-parser-combinators/issues/550

**Comment by catap** [comment]
@cheeseng CI is failing on scala-js. Anyway, I assume that you have already figured it out because main branch has updated version as well.

So, feel free to harvest from this PR support of SN-0.5.0 and dependencies.

But all of them should be used with modern enoug Scala.js.

**Comment by cheeseng** [comment]
@catap Fyi I am trying this now.

**Comment by cheeseng** [comment]
@catap I have staged 3.2.18 build for scala-native 0.5, and submitted the following PR based on this PR: 

https://github.com/scalatest/scalatest/pull/2322

@bvenners shall have the pleasure to release it.

Cheers.

**Comment by catap** [comment]
Closed in favour of #2322

**Pull request closed without merging**